x/stakeibc/keeper/validator_selection.go

```diff
@@ -4,60 +4,255 @@ import (
+	}
+	if hostZone.DelegationIcaAddress == "" {
+		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", chainId)
+	}
+
+	// Get the difference between the actual and expected validator delegations
+	valDeltaList, err := k.GetValidatorDelegationDifferences(ctx, hostZone)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to get validator deltas for host zone %s", chainId)
+	}
+
+	msgs, rebalacings := k.GetRebalanceICAMessages(hostZone, valDeltaList, numRebalance)
+
+	// marshall the callback
+	rebalanceCallback := types.RebalanceCallback{
+		HostZoneId:   hostZone.ChainId,
+		Rebalancings: rebalacings,
+	}
+	rebalanceCallbackBz, err := k.MarshalRebalanceCallbackArgs(ctx, rebalanceCallback)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to marshal rebalance callback args")
+	}
+
+	// Submit the rebalance ICA
+	_, err = k.SubmitTxsStrideEpoch(
+		ctx,
+		hostZone.ConnectionId,
+		msgs,
+		types.ICAAccountType_DELEGATION,
+		ICACallbackID_Rebalance,
+		rebalanceCallbackBz,
+	)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Failed to SubmitTxs for %s, messages: %+v", hostZone.ChainId, msgs)
+	}
+
+	return nil
+}
+
+// Given a list of target delegation changes, builds the individual re-delegation messages by redelegating
+// from overweight validators to underweight validators
+// Returns the list of messages and the callback data for the ICA
+func (k Keeper) GetRebalanceICAMessages(
+	hostZone types.HostZone,
+	validatorDeltas []RebalanceValidatorDelegationChange,
+	numRebalance uint64,
+) (msgs []sdk.Msg, rebalancings []*types.Rebalancing) {
+	// Sort the list of delegation changes by the size of the change
+	lessFunc := func(i, j int) bool {
+		if !validatorDeltas[i].Delta.Equal(validatorDeltas[j].Delta) {
+			return validatorDeltas[i].Delta.LT(validatorDeltas[j].Delta)
+		}
+		// use name as a tie breaker if deltas are equal
+		return validatorDeltas[i].ValidatorAddress < validatorDeltas[j].ValidatorAddress
+	}
+	sort.SliceStable(validatorDeltas, lessFunc)
+
+	// QUESTION: Does anyone have a preference on the naming convention here? A couple options:
+	//  * Overweight/Underweight (current)
+	//  * Surplus/Deficit
+	//  * Source/Destination
+	//  * Giver/Receiver
+	// etc.
+
+	// Pair overweight and underweight validators, with a redelegation from the overweight
+	// validator to the underweight one
+	// The list is sorted with the overweight validators (who should lose stake) at index 0
+	//   and the underweight validators (who should gain stake) at index N-1
+	// The overweight validator's have a negative delta and the underweight validators have a positive delta
+	overWeightIndex := 0
+	underWeightIndex := len(validatorDeltas) - 1
+	for i := uint64(1); i <= numRebalance; i++ {
+		// underweight Elem is positive, overweight Elem is negative
+		underWeightValidator := validatorDeltas[underWeightIndex]
```

Comment:
I'm definitely on board with adding redundant checks if they help readability! But just to make sure I'm following, you're thinking something like:
```go
if overWeightIndex == underWeightIndex {
  break
}
// or 
if overWeightValidator.Delta < underWeightValidator.Delta {
  break
}
```

Considering that that first case doesn't happen every time, and the second case assumes a knowledge of the delta signs (which is a bit hard to remember), I personally feel like they both make the code a little harder to follow, but will definitely defer to you since you're coming from the reviewer/reader perspective atm!

Reply:
Oh I was thinking something slightly different, more like
`
if overWeightIndex <= underWeightIndex {
  break
}
`
I think we always want to break if the two indices have flipped!

Reply:
just added!